pydantic/_internal/_core_metadata.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations as _annotations
+
+import typing
+import warnings
+from typing import Any
+
+import typing_extensions
+from pydantic_core import CoreSchema
+from pydantic_core.core_schema import TypedDictField
+
+from ..json_schema_misc import JsonSchemaMisc
+from ._typing_extra import EllipsisType
+
+_UPDATE_CORE_SCHEMA_FIELD = 'pydantic_update_core_schema'
+_JSON_SCHEMA_MISC_FIELD = 'pydantic_json_schema_misc'
+
+_CoreMetadata = typing.Dict[str, Any]
+
+
+class UpdateCoreSchemaCallable(typing_extensions.Protocol):
+    def __call__(self, schema: CoreSchema, **kwargs: Any) -> None:
+        ...
+
+
+class CoreMetadataHandler:
+    """
+    Because the metadata field in pydantic_core is of type `Any`, we can't assume much about its contents.
+
+    This class is used to interact with the metadata field on a CoreSchema object in a consistent
+    way throughout pydantic.
+    """
+
+    def __init__(self, schema: CoreSchema | TypedDictField):
+        self.schema = schema
+
+        metadata = schema.get('metadata')
+        if metadata is None:
+            schema['metadata'] = {}
+        elif not isinstance(metadata, dict):
+            raise ValueError(f'CoreSchema metadata should be a dict; got {metadata!r}.')
+
+    @property
+    def metadata(self) -> _CoreMetadata:
+        metadata = self.schema.get('metadata')
+        if metadata is None:
+            self.schema['metadata'] = metadata = {}
+        if not isinstance(metadata, dict):
+            raise ValueError(f'CoreSchema metadata should be a dict; got {metadata!r}.')
```

Comment:
After playing around with the implications a bit, I felt like it was significantly cleaner to just raise errors if the CoreSchema metadata isn't at least a dict. Too many places things would get mishandled otherwise, and I think it's reasonable. I would vote that we change the type hint in `pydantic_core` from `Any` to `dict[str, Any] | None` (or even just `dict[str, Any]`, honestly, though if retaining None as a choice is more performant that's fine with me.)